nwg_panel/tools.py

```diff
@@ -601,79 +601,68 @@ def player_metadata():
-            image.set_from_pixbuf(pixbuf)
-        except:
-            pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(
-                os.path.join(get_config_dir(), "icons_light/icon-missing.svg"), icon_size, icon_size)
-            image.set_from_pixbuf(pixbuf)
-    else:
-        icon_theme = Gtk.IconTheme.get_default()
-        if icons_path:
-            path = "{}/{}.svg".format(icons_path, icon_name)
-            try:
-                pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(path, icon_size, icon_size)
-                if image:
-                    image.set_from_pixbuf(pixbuf)
-            except:
-                try:
-                    pixbuf = icon_theme.load_icon(icon_name, icon_size, Gtk.IconLookupFlags.FORCE_SIZE)
-                    if image:
-                        image.set_from_pixbuf(pixbuf)
-                except:
-                    pass
-        else:
-            try:
-                pixbuf = icon_theme.load_icon(icon_name, icon_size, Gtk.IconLookupFlags.FORCE_SIZE)
-            except:
-                try:
-                    pixbuf = icon_theme.load_icon(icon_name.lower(), icon_size, Gtk.IconLookupFlags.FORCE_SIZE)
-                except:
-                    path = os.path.join(get_config_dir(), "icons_light/icon-missing.svg")
-                    pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(path, icon_size, icon_size)
-            if image:
-                image.set_from_pixbuf(pixbuf)
+def update_image(image, icon_name, icon_size, icons_path="", fallback=True):
+    scale = image.get_scale_factor()
+    icon_size *= scale
+    pixbuf = create_pixbuf(icon_name, icon_size, icons_path, fallback)
+    surface = Gdk.cairo_surface_create_from_pixbuf(pixbuf, scale, image.get_window())
+    image.set_from_surface(surface)
 
 
-def create_pixbuf(icon_name, icon_size, icons_path=""):
-    # In case a full path was given
-    if icon_name.startswith("/"):
-        try:
+def update_image_fallback_desktop(image, icon_name, icon_size, icons_path):
+    try:
+        # This should work if your icon theme provides the icon, or if it's placed in /usr/share/pixmaps
+        update_image(image, icon_name, icon_size, fallback=False)
+    except:
+        # If the above fails, let's search .desktop files to find the icon name
+        icon_from_desktop = get_icon_name(icon_name)
+        if icon_from_desktop:
+            # trim extension, if given and the definition is not a path
+            if "/" not in icon_from_desktop:
+                icon_from_desktop = os.path.splitext(icon_from_desktop)[0]
+
+            update_image(image, icon_from_desktop, icon_size, icons_path)
+
+
+def update_gtk_entry(entry, icon_pos, icon_name, icon_size, icons_path=""):
+    scale = entry.get_scale_factor()
+    icon_size *= scale
+    pixbuf = create_pixbuf(icon_name, icon_size, icons_path)
+    entry.set_icon_from_pixbuf(icon_pos, pixbuf)
+
+
+def create_pixbuf(icon_name, icon_size, icons_path="", fallback=True):
```

Comment:
Summary of changes:
1. Observing the original `create_pixbuf` always fallbacks to `icons_light/icon-missing.svg`, it's moved to the outmost try-except block so we no longer need to duplicate it 3 times.
2. `if icon_name.startswith("/")`, then load from the absolute path (same as before). Otherwise, there are 3 routines in order if `icons_path` supplied:
    1.  If `icons_path` supplied, load from `icons_path/icon_name.svg`
    2. load icon_name from icon theme
    3. convert icon_name to lowercase and load from icon theme

    and if `icons_path` is not supplied, skip i. In short, ii. and iii. are executed when 
Previously, ii. and iii. were duplicated in i.'s if and else blocks. This commit avoids duplication by putting ii. and iii. in i.'s except block.

The added `fallback` option raises an exception when the icon is not found, rather than fallback to `icon-missing.svg`. This enables the caller to do custom error handling.